packages/devtools/lib/src/timeline/timeline_protocol.dart

```diff
@@ -112,102 +210,149 @@ class TimelineData {
     return '${event.name}-${event.id}';
   }
 
-  void _handleDurationBeginEvent(TraceEvent event) {
-    final e = TimelineEvent(
+  void _handleDurationBeginEvent(TraceEvent event, int orderId) {
+    final current = currentEventNodes[event.type.index];
+    if (current == null &&
+        !(event.name.contains('VSYNC') ||
+            event.name.contains('GPURasterizer::Draw'))) {
+      return;
+    }
+
+    final timelineEvent = TimelineEvent(
       event.name,
       event.timestampMicros,
       event.type,
+      orderId,
+      event.json,
     );
 
-    if (event.isCpuEvent) {
-      if (_cpuEventNode != null) {
-        _cpuEventNode.addChild(e);
-        _cpuEventNode = e;
-      }
-      // Do not add MessageLoop::RunExpiredTasks events to a null stack. These
-      // events will either a) start outside of our frame start time, or b)
-      // parent irrelevant events - neither of which we want.
-      else if (!event.name.contains('MessageLoop::RunExpiredTasks')) {
-        _cpuEventNode = e;
-      }
-    } else if (event.isGpuEvent) {
-      if (_gpuEventNode != null) {
-        _gpuEventNode.addChild(e);
-        _gpuEventNode = e;
-      }
-      // Do not add MessageLoop::RunExpiredTasks events to a null stack. A
-      // single MessageLoop::RunExpiredTasks event can parent multiple
-      // PipelineConsume event flows, and we want to consider each
-      // PipelineConsume flow to be its own event. This event can also parent
-      // irrelevant events that we do not want to track.
-      else if (!event.name.contains('MessageLoop::RunExpiredTasks')) {
-        _gpuEventNode = e;
-      }
+    if (current != null) {
+      current.addChild(timelineEvent);
     }
+    currentEventNodes[event.type.index] = timelineEvent;
   }
 
   void _handleDurationEndEvent(TraceEvent event) {
-    TimelineEvent current;
-    if (event.isCpuEvent && _cpuEventNode != null) {
-      _cpuEventNode.endTime = event.timestampMicros;
-      current = _cpuEventNode;
-
-      // Since this event is complete, move back up the stack.
-      _cpuEventNode = _cpuEventNode.parent;
-      if (_cpuEventNode == null) {
-        _maybeAddEvent(current);
+    TimelineEvent current = currentEventNodes[event.type.index];
+
+    if (current == null) return;
+
+    // If the names of [event] and [current] do not match, our event nesting is
+    // off balance due to duplicate events from the engine. Balance the tree so
+    // we can continue processing trace events for [current].
+    if (event.name != current.name) {
+      if (event.json.toString() ==
```

Comment:
is this really an error if the duration of the two events is zero? We should dig into what this case means on the VM side.
in any case, we should compare based on equality not based on calling toString.